src/merkle/tiered_smt/nodes.rs

```diff
@@ -0,0 +1,365 @@
+use super::{
+    get_index_tier, get_key_prefix, is_tiered_node, BTreeMap, BTreeSet, EmptySubtreeRoots,
+    InnerNodeInfo, MerkleError, MerklePath, NodeIndex, Rpo256, RpoDigest, Vec,
+};
+
+// CONSTANTS
+// ================================================================================================
+
+/// The number of levels between tiers.
+const TIER_SIZE: u8 = super::TieredSmt::TIER_SIZE;
+
+/// Depths at which leaves can exist in a tiered SMT.
+const TIER_DEPTHS: [u8; 4] = super::TieredSmt::TIER_DEPTHS;
+
+/// Maximum node depth. This is also the bottom tier of the tree.
+const MAX_DEPTH: u8 = super::TieredSmt::MAX_DEPTH;
+
+// NODE STORE
+// ================================================================================================
+
+/// A store of nodes for a Tiered Sparse Merkle tree.
+///
+/// The store contains information about all nodes as well as information about which of the nodes
+/// represent leaf nodes in a Tiered Sparse Merkle tree.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct NodeStore {
+    nodes: BTreeMap<NodeIndex, RpoDigest>,
+    upper_leaves: BTreeSet<NodeIndex>,
+    bottom_leaves: BTreeSet<u64>,
```

Comment:
question: why are the leaves split into upper and bottom? I couldn't figure it out from reading the code.

Reply:
request: it would be nice to write a comment with the invariants of upper/bottom_leaves. From what I gathered it is roughly this:

> Sets used to determine the position of the leaves in the tree. A leaf can never be a parent of another leaf.

Reply:
> why are the leaves split into upper and bottom? I couldn't figure it out from reading the code.

This is mainly done to speed up iteration over bottom leaves (e.g., via `bottom_leaves()` iterator). But I also think that it may make sense, for efficiency reasons, to split this out even further. For example, we could have a separate variable to track leaves at depth 16. This could be represented not as a `BTreeSet` but as a single vector of bits which should make lookups rather fast.

But we should have a proper set of benchmarks before we do these types of optimizations.

Reply:
> it would be nice to write a comment with the invariants of upper/bottom_leaves. From what I gathered it is roughly this:
> 
> > Sets used to determine the position of the leaves in the tree. A leaf can never be a parent of another leaf.

I updated the comment to include something similar to the first sentence but skipped the definition of what is a leaf because I think this is an accepted definition of a leaf anyway.